src/loaders/OMETiffLoader.js

```
@@ -0,0 +1,322 @@
+  }
+}
+
+function flip32(info) {
+  const flipper = new Uint8Array(info.buffer, info.byteOffset, info.length * 4);
+  const len = flipper.length;
+  for (let i = 0; i < len; i += 4) {
+    let t = flipper[i];
+    flipper[i] = flipper[i + 3];
+    flipper[i + 3] = t;
+    t = flipper[i + 1];
+    flipper[i + 1] = flipper[i + 2];
+    flipper[i + 2] = t;
+  }
+}
+
+function flip64(info) {
+  const flipper = new Uint8Array(info.buffer, info.byteOffset, info.length * 8);
+  const len = flipper.length;
+  for (let i = 0; i < len; i += 8) {
+    let t = flipper[i];
+    flipper[i] = flipper[i + 7];
+    flipper[i + 7] = t;
+    t = flipper[i + 1];
+    flipper[i + 1] = flipper[i + 6];
+    flipper[i + 6] = t;
+    t = flipper[i + 2];
+    flipper[i + 2] = flipper[i + 5];
+    flipper[i + 5] = t;
+    t = flipper[i + 3];
+    flipper[i + 3] = flipper[i + 4];
+    flipper[i + 4] = t;
+  }
+}
+
+function flipEndianness(arr) {
+  switch (arr.BYTES_PER_ELEMENT) {
+    case 1:
+      // no op
+      return;
+    case 2:
+      flip16(arr);
+      break;
+    case 4:
+      flip32(arr);
+      break;
+    case 8:
+      flip64(arr);
+      break;
+    default:
+      throw new Error('Invalid input');
+  }
+}
+
+/**
+ * This class serves as a wrapper for fetching tiff data from a file server.
+ * */
+export default class OMETiffLoader {
+  constructor(tiff, poolOrDecoder, firstImage, omexmlString, offsets) {
+    this.poolOrDecoder = poolOrDecoder;
+    this.tiff = tiff;
+    this.type = 'ome-tiff';
+    // get first image's description, which contains OMEXML
+    this.omexml = new OMEXML(omexmlString);
+    this.offsets = offsets || [];
+    this.channelNames = this.omexml.getChannelNames();
+    this.width = this.omexml.SizeX;
+    this.height = this.omexml.SizeY;
+    this.tileSize = firstImage.getTileWidth();
+    const { SubIFDs } = firstImage.fileDirectory;
+    this.numLevels =
+      this.omexml.getNumberOfImages() || (SubIFDs && SubIFDs.length);
```

could the omexml return `null` or `undefined`?

---

I don't think so unless something in the constructor of `OMEXML` breaks. In that case, it's probably ok to let things fail loudly.